### PR TITLE
Converted naiveBayesBowDemo to python

### LINKED ISSUE #199: Convert naiveBayesBowDemo to python
This is figure number 9.7 in the book. its matlab code is [here](https://github.com/probml/pmtk3/blob/master/demos/naiveBayesBowDemo.m)

### PR DESCRIPTION
**Pull request by ashishpapanai**
Convert [naiveBayesBowDemo](https://github.com/probml/pmtk3/blob/master/demos/naiveBayesBowDemo.m) to python. 
Closes #199 
Plots Generated So far: 
Figure_1
Figure_2

### CODE REVIEW

**Comment by ashishpapanai** [comment]
@murphyk , I've converted the code to create plots as in figure 9.7. 
I am having some issues with the word LatexTable.

**Comment by ashishpapanai** [comment]
@mjsML Please review the PR.

**Review by mjsML — CHANGES_REQUESTED** [review]
Can I ask you to please make sure 
- That the labels are the same as the book.
- That the script is error free.
- That unused code is removed.
- Please make sure to use colab and our setup, visit [http://figures.probml.ai/book1/9.7](http://figures.probml.ai/book1/9.7) and run the setup cell (should clone our repo to the remote colab machine) then run your code (an error regarding the data path will come up).

**Comment by ashishpapanai** [comment]
Hey @mjs: 

- The data path is rectified. 
- Unnecessary code from the script is removed. 
- The script is now error-free. 

Please suggest if there are more issues/ changes required.

**Review by mjsML — CHANGES_REQUESTED** [review]
- What is the use of the if statement in line 15? if there is no use please remove it.
- use the path like in colab e.g. line 16 `data = read_mat('../data/XwindowsDocData.mat')`
- same for the save figure line 38 ` plt.savefig(r'../figures/'+fileName)`
- Please test in this [colab](https://colab.research.google.com/github/probml/pyprobml/blob/master/book1/figures/chapter9_figures.ipynb#9.7) , remove the show image functions and paste your code and make sure there are no errors.

**Comment by ashishpapanai** [comment]
> * What is the use of the if statement in line 15? if there is no use please remove it. 
> * use the path like in colab e.g. line 16 `data = read_mat('../data/XwindowsDocData.mat')`
> * same for the save figure line 38 ` plt.savefig(r'../figures/'+fileName)`
> * Please test in this [colab](https://colab.research.google.com/github/probml/pyprobml/blob/master/book1/figures/chapter9_figures.ipynb#9.7) , remove the show image functions and paste your code and make sure there are no errors.

1. Removed Line 15 as it was useless and should have been removed with other un-necessary lines. 
2. Changes the path for colab, this path won't work on windows machine as the path is relative, the previous path written by me was working fine on windows but was throwing an error on colab [Fixed]
3. Done
4.  The image generated on colab is different than the image generated locally on my device, I've attached the images in the body of my PR, please refer to that for review.

**Comment by murphyk** [comment]
@ashishpapanai When I run your code in a colab, I get the image below. It does not have a spike of prob=1.0 for the word "subject". I think you should not use Gaussian Naive Bayes, since data is binary.
download

**Comment by ashishpapanai** [comment]
@murphyk Trying Bernoulli Naive Bias, but what about theta in Bernoulli Naive Bias?

**Comment by murphyk** [comment]
Theta in my matlab count is just the empirical matrix of class conditional frequencies.
It seems that in sklearn the relevant attribute is feature_log_prob_

**Comment by ashishpapanai** [comment]
The code is running fine in the Local Environment, But it is giving anomalous output in Colab. I've replaced Gaussian Naive Bayes with Bernoulli Naive Bayes which is used for binary data but it turns out futile in Colab. 

Figure_12

**Comment by ashishpapanai** [comment]
I think that model training is causing this discrepancy, as the misclassification rate on training is around 8.33 on my PC and for the test, it is around 18%. But in colab, the misclassification rate on training is around 24% and on the test, it is around 44%.
Screenshot (692)
Screenshot (691)

**Comment by mjsML** [comment]
@ashishpapanai I run ubuntu, @murphyk runs macos and you run windows 🙂 colab is a common denominator, and it eases reproducibility and testing. This is why we kindly ask that everything is done/ verified in colab. I suspect this might be a library versioning issue (and this is why colab is a better choice for a "sounding board")

**Comment by ashishpapanai** [comment]
@mjsML, Colab is running scikit-learn version 0.22.2.post1, were as on my local machine I have 0.24.1. 
I'll give it a try on Ubuntu as well as the algorithm seems to be correct to me. Should I convert the Matlab implementation of Naive Bayes to python from scratch instead of using the scikit-learn version of BernoulliNB?

**Comment by mjsML** [comment]
> @mjsML, Colab is running scikit-learn version 0.22.2.post1, were as on my local machine I have 0.24.1.
> I'll give it a try on Ubuntu as well as the algorithm seems to be correct to me. Should I convert the Matlab implementation of Naive Bayes to python from scratch instead of using the scikit-learn version of BernoulliNB?

Did you try to upgrade the colab version by installing the latest version of scikit-learn and it produced the same result as your local machine?
The point here is we should all use colab and avoid the typical "it works on my machine" discussions 🙂

**Comment by ashishpapanai** [comment]
I'm sorry, I didn't mean to get into that conversation.
I wanted to convey that even after updating scikit-learn it's not working.
So should I implement the 3 supporting functions used to create a Naive
Bayes model and discard using scikit-learn for it?

**Comment by mjsML** [comment]
I don't think it's the case, I believe the issue is the model is fitted twice (which is not supposed to happen).
try
```python

model = BernoulliNB()

model.fit(Xtrain, ytrain)

ypred_train = model.predict(Xtrain)

err_train = np.mean(zero_one_loss(ytrain, ypred_train))

ypred_test = model.predict(Xtest)

err_test = np.mean(zero_one_loss(ytest, ypred_test))
```

This gave me on colab the same error rates you have on your local machine:
image


Training the model twice probably overfitted it.

Also the figure title needs to be :
```python 
plt.title('p(xj=1|y='+str(i+1)+')')
``` 
as per the graph label in the book.
We still don't have the same graph though 😅

**Comment by mjsML** [comment]
GaussianNB with theta_ gives this on colab FYI:

image

**Comment by ashishpapanai** [comment]
That's the main issue as the data is binary, I changed GaussianNB to BernoulliNB but it gave exactly the same output as GaussianNB.
I think the only option left is to convert the following functions: [NaiveBiasFir](https://github.com/probml/pmtk3/blob/master/toolbox/SupervisedModels/naiveBayes/naiveBayesFit.m), [naiveBayesPredict](https://github.com/probml/pmtk3/blob/master/toolbox/SupervisedModels/naiveBayes/naiveBayesPredict.m) and other auxiliary functions to python.

**Comment by ashishpapanai** [comment]
@mjsML, Tried on ubuntu as well, I guess the issue is with colab. If possible, could you please try to run the version on your local machine, If the output is the same as colab, I'll reinvent the wheel by coding Naive Bias from scratch.

**Comment by ashishpapanai** [comment]
@mjsML @murphyk The issue is with Colab Kernel, I tried testing code on Kaggle Kernal and it works perfectly fine. [Kaggle Output](https://www.kaggle.com/ashishpapanai/notebooka7e7260202)

**Comment by mjsML** [comment]
The plot thickens (pun intended) 😄 

run this in  kaggle:
```python
print(np.exp(model.feature_log_prob_)[i, :].tolist())
``` 
and copy the print result
now run this in colab:
```python

Colab_prob = np.exp(model.feature_log_prob_)[i, :].tolist()
Kaggle_prob = #Paste your kaggle list result
for i,j in zip(Colab_prob,Kaggle_prob):
  if i! = j:
    print(True)
index = 106
print(Colab_prob[index],Kaggle_prob[index])
Colab_prob[index] == Kaggle_prob[index]
```
!!

**Comment by murphyk** [comment]
If this is a problem with colab kernel, this bug needs to be reported to them. 
If it is a problem with sklearn (seems unlikely), this bug needs to be reported to them. 
You could obviously implement the functions yourself, but fixing this bug seems more important :)

**Comment by ashishpapanai** [comment]
@mjsML I tried the test you've mentioned, It turns out that `np.exp(model.feature_log_prob_)[i, :]` generated on colab and kaggle are exactly the same. 
The data is exactly the same on Kaggle and on Colab but the plots are different. 

1. I've tested Colab's  `np.exp(model.feature_log_prob_)` on Kaggle and It's the same as Colab and gave the required output plot.  
Screenshot (694)
Screenshot (696)
Screenshot (695)

2. Tested Kaggle's `np.exp(model.feature_log_prob_)` on Colab and it gave the anomalous plot. 
Screenshot (697)

**Comment by mjsML** [comment]
@ashishpapanai , I tested also tinkering with all the versions combinations you'd think of and even added conda to colab to match the entire stack, colab just refuses to output the right plot (although it's a simple bar chart!) ... I'm honestly baffled by this. We probably should file a proper issue with the colab Team later like @murphyk suggested.
I would park this for now though.

**Comment by ashishpapanai** [comment]
So this won't be merged for now? ☹

**Comment by mjsML** [comment]
@ashishpapanai I understand this is frustrating.
But technically the code doesn't achieve its purpose hence I wouldn't merge it. Point is maybe the colab guys will tell us oh ya this is a well known issue and you need to use super bar from foo.lib .
Try posting on stackoverflow with the colab tag, some of the colab engineers monitor those.
Link the stackoverflow discussion here so we can follow it. Although I doubt this is a "user error". I strongly believe this might be a deeper issue than the simple pass two arrays to a graphing library to build a bar chart.

**Comment by animesh-007** [comment]
Hello @ashishpapanai @murphyk @mjsML 
So I was trying to figure out this issue. I found that this issue was mainly due to dependencies version differences. I have also provided my colab and Kaggle notebook for reference.

- [Colab Notebook](https://colab.research.google.com/drive/1kSgUxN4lPd5YA1is4WWGVGEY4kZnao9L?usp=sharing)

- [Kaggle Notebook](https://www.kaggle.com/animesh2099/naivebayesbowdemo-ipynb)

**Comment by ashishpapanai** [comment]
@animesh-007 Tried asking the reason for this on [StackOverflow](https://stackoverflow.com/questions/66645378/different-plot-of-the-same-data-is-being-generated-in-google-colab-and-kaggle), Got 2 downvotes and 0 Solution. 
Anyways, Is there any fix for this that will create the required plots in colab?

**Comment by animesh-007** [comment]
> @animesh-007 Tried asking the reason for this on [StackOverflow](https://stackoverflow.com/questions/66645378/different-plot-of-the-same-data-is-being-generated-in-google-colab-and-kaggle), Got 2 downvotes and 0 Solution.
> Anyways, Is there any fix for this that will create the required plots in colab?

You can write your code inside a python file and run it using the following command `!python [name_of_file].py`. It will generate the correct plot in this case. You can also write your entire code like this 
```
%%bash
python
code
```
This will also generate the correct plot. These two methods will generate the correct plot inside colab. For reference, you can also check my colab notebook.

**Comment by ashishpapanai** [comment]
@animesh-007 That's a platform-dependent solution. As suggested by @mjsML a platform-independent solution is required to close the issue.

**Comment by animesh-007** [comment]
> @animesh-007 That's a platform-dependent solution. As suggested by @mjsML a platform-independent solution is required to close the issue.

I have tried this on colab and it worked perfectly there.😕

**Comment by ashishpapanai** [comment]
> > @animesh-007 That's a platform-dependent solution. As suggested by @mjsML a platform-independent solution is required to close the issue.
> 
> I have tried this on colab and it worked perfectly there.😕

No, I meant if there's a single fix to run the code flawlessly as both, a python file and a colab cell. 
The colab notebook that you've attached works fine.

**Comment by murphyk** [comment]
I didn't follow the whole discussion about colab vs kaggle, but I did try running the colab from @animesh-007, which uses BernoulliNB. This works as intended and exactly matches the matlab code:
```
logprobs = model.feature_log_prob_
probs = np.exp(logprobs)
print(probs[:,107-1]) # produces [0.99778761 0.99778761]
```
The problem was just in the printing! 
The spike of 1.0 at feature 107 is so thin that it does not show up unless you make the plot bigger!
download
I will close this issue and submit a simplified version of the code.
Thank you all for working so hard on this!

**Comment by murphyk** [comment]
https://github.com/probml/pyprobml/blob/master/scripts/naiveBayesBowDemo.py

**Comment by murphyk** [comment]
I gave you both credit in the script :)
